File: tests/test_lru_state.py
import unittest
import sys
sys.path.append('../src')

from LRU_state import *

class TestLruState(unittest.TestCase):
    def test_integration(self):
        lru_state = LRU_state(4, 0)

        s0 = '00000'
        s1 = '00001'
        s2 = '00010'
        s3 = '00011'
        s4 = '00100'
        states = [s0, s1, s2, s0, s3, s4]

        results = []
        for state in states:
            next_state = lru_state.get_closest_state(state)
            if len(next_state) > 0:
                lru_state.enqueue(next_state)
            else:
                lru_state.enqueue(state)
            results.append(str(lru_state))

        actual_result = '\n'.join(results)
        expected_result = """odict_items([('00000', 1)])
odict_items([('00000', 1), ('00001', 1)])
odict_items([('00000', 1), ('00001', 1), ('00010', 1)])
odict_items([('00001', 1), ('00010', 1), ('00000', 2)])
odict_items([('00001', 1), ('00010', 1), ('00000', 2), ('00011', 1)])
odict_items([('00010', 1), ('00000', 2), ('00011', 1), ('00100', 1)])"""

        self.assertEqual(actual_result, expected_result)

if __name__ == '__main__':
    unittest.main()

File: src/cpp/test/test_lru_state.cpp
#include <assert.h>
#include <iostream>
#include "lru_state.h"

int main() {
    lru_state* state_queue = new lru_state(4, 0);


    set<int> n0 = {0};
    set<int> n1 = {4};
    set<int> n2 = {3};
    set<int> n3 = {3, 4};
    set<int> n4 = {2};
    
    vector<set<int>> states = {n0, n1, n2, n0,n3, n4};
    
    vector<string> expected_result = {
        "0:1->",
        "4:1->0:1->",
        "3:1->4:1->0:1->",
        "0:2->3:1->4:1->",
        "3,4:1->0:2->3:1->4:1->",
        "2:1->3,4:1->0:2->3:1->"
    };
    
    set<int> ids_to_exclude;

    for (int i = 0; i < states.size(); i++) {
        set<int> closest_state = state_queue->get_closest_state(states[i], ids_to_exclude);
        state_queue->enqueue(states[i]);

        string cur_state_queue = state_queue->to_string();
        // std::cout << cur_state_queue << std::endl;
        assert(cur_state_queue == expected_result[i]);
    }
    
    return 0;
}
